nuxeo-core/nuxeo-core-io/src/main/java/org/nuxeo/ecm/core/io/marshallers/csv/DocumentModelCSVWriter.java

```diff
@@ -0,0 +1,137 @@
+import java.util.Arrays;
+import java.util.Calendar;
+import java.util.Collections;
+import java.util.List;
+
+import javax.inject.Inject;
+import javax.ws.rs.core.MediaType;
+
+import org.joda.time.DateTime;
+import org.joda.time.format.ISODateTimeFormat;
+import org.nuxeo.ecm.core.api.DocumentModel;
+import org.nuxeo.ecm.core.api.Lock;
+import org.nuxeo.ecm.core.api.model.Property;
+import org.nuxeo.ecm.core.io.registry.MarshallerRegistry;
+import org.nuxeo.ecm.core.io.registry.Writer;
+import org.nuxeo.ecm.core.io.registry.reflect.Setup;
+import org.nuxeo.ecm.core.schema.SchemaManager;
+import org.nuxeo.ecm.core.schema.types.Field;
+import org.nuxeo.ecm.core.schema.types.Schema;
+import org.nuxeo.ecm.core.schema.utils.DateParser;
+
+/**
+ * @since 10.3
+ */
+@Setup(mode = SINGLETON, priority = REFERENCE)
+public class DocumentModelCSVWriter extends AbstractCSVWriter<DocumentModel> {
+
+    public DocumentModelCSVWriter() {
+        super(DocumentModel.class);
+    }
+
+    @Inject
+    protected SchemaManager schemaManager;
+
+    @Inject
+    protected MarshallerRegistry registry;
+
+    @Override
+    public void write(DocumentModel entity, Class<?> clazz, Type genericType, MediaType mediatype, OutputStream out)
+            throws IOException {
+        writeSystem(out, entity);
+        // schemas are sorted for reproducibility
+        List<String> schemas = new ArrayList<>(Arrays.asList(entity.getSchemas()));
+        Collections.sort(schemas);
+        for (int i = 0; i < schemas.size(); i++) {
+            writeSchema(out, entity, schemas.get(i));
+            if (i < schemas.size() - 1) {
+                writeSeparator(out);
+            }
+        }
+    }
+
+    /**
+     * CSV must have consistent column number, null and empy values have to be written.
+     */
+    protected void writeSystem(OutputStream out, DocumentModel doc) throws IOException {
+        writeWithSeparator(out, doc.getId());
+        writeWithSeparator(out, doc.getPathAsString());
+        writeWithSeparator(out, doc.getRepositoryName());
+        writeWithSeparator(out, doc.getType());
+        writeWithSeparator(out, doc.getRef() != null ? doc.getCurrentLifeCycleState() : null);
+        writeWithSeparator(out, doc.getTitle());
+        writeWithSeparator(out, doc.getParentRef() != null ? doc.getParentRef().toString() : null);
+        writeWithSeparator(out, Boolean.toString(doc.isTrashed()));
+        writeWithSeparator(out, Boolean.toString(doc.isVersion()));
+        writeWithSeparator(out, Boolean.toString(doc.isProxy()));
+        writeWithSeparator(out, Boolean.toString(doc.isCheckedOut()));
+        writeWithSeparator(out, doc.getChangeToken());
+        writeWithSeparator(out, doc.getVersionLabel());
+        Lock lock = doc.getLockInfo();
+        if (lock != null) {
+            writeWithSeparator(out, lock.getOwner());
+            writeWithSeparator(out, ISODateTimeFormat.dateTime().print(new DateTime(lock.getCreated())));
```

Comment:
Don't use joda, `java.time` could do the same thing.

Reply:
Furthermore, maybe `writeWithSeparator` accepting a `Date` or a `Calendar` in abstract implementation could be a useful thing (as we may always serialize date in the same format).